Add expiration date to share by link view

MorrisJobke opened the pull request:
![bildschirmfoto von 2015-10-01 15-59-07](https://cloud.githubusercontent.com/assets/245432/10222652/68d4afe6-6855-11e5-8995-a4eb7a5550c1.png)

Talked about this with @felixboehm 

Maybe @PVince81 can have a look at the actual implementation.

MorrisJobke committed:
Add expiration date to share by link view

karlitschek commented:
nice small improvement :+1:

PVince81 commented on apps/files_sharing/js/sharedfilelist.js:
```diff
@@ -79,6 +79,38 @@
+				// difference in days multiplied by 5 - brightest shade for expiry dates in more than 32 days (160/5)
+				var modifiedColor = Math.round((expirationTimestamp - (new Date()).getTime()) / 1000 / 60 / 60 / 24 * 5);
+				// ensure that the brightest color is still readable
+				if (modifiedColor >= '160') {
```
'160' as string ?!

MorrisJobke replied:
I just copied from `filelist.js` :see_no_evil:

PVince81 commented on apps/files_sharing/js/sharedfilelist.js:
```diff
@@ -79,6 +79,38 @@
+					formatted = OC.Util.formatDate(expirationTimestamp);
+					text = OC.Util.relativeModifiedDate(expirationTimestamp);
+				} else {
+					formatted = t('files_sharing', 'No expiration date set');
```
Not sure if we want to show something here, @jancborchardt what do you think ? Maybe a shorter message.

phil-davis replied:
'Never' ?

MorrisJobke replied:
As you can see this is only the tooltip. (see third row in the screenshot above)

I can also permanently show a "Never" instead of the empty cell.

jancborchardt replied:
I’d say »Never« or even nothing at all is better.

MorrisJobke replied:
> I’d say »Never« or even nothing at all is better.

The cell itself is empty. This is only the tooltip. Should the tooltip be empty and the cell should say "Never"? Or what is the way to go here?

jancborchardt replied:
Leave it empty, and the tooltip should say »No expiration date set«. When a date is set, the tooltip is for the exact date I assume (whereas the displayed date is a relative date, like in the file list).

PVince81 commented on apps/files_sharing/js/sharedfilelist.js:
```diff
@@ -79,6 +79,38 @@
+						"class": "modified",
+						"title": formatted,
+						"style": 'color:rgb(' + modifiedColor + ',' + modifiedColor + ',' + modifiedColor + ')'
+					}).text(text)
```
One day we'll use handlebars. Pretty sure you copy-pasted this from somewhere :wink: 
Looks fine.

MorrisJobke replied:
Yes. I copy pasted this from `fileslist.js`

MorrisJobke committed:
add "Never" if expiration date is not set

MorrisJobke commented:
@phil-davis @PVince81 I addressed your comments:

![bildschirmfoto von 2015-10-02 08-44-34](https://cloud.githubusercontent.com/assets/245432/10240905/f4b1203a-68e1-11e5-82b9-4f73a0432c8c.png)

PVince81 commented:
will review/test after the 8.2 rush...

MorrisJobke commented:
Another problem that I noticed during working on this: Sorting is handled in the FileList in a generic way. But it has one combined attribute to "How to sort" (time based) and "What to sort" (share time). Currently it only has sort by "mtime" which is sort time based the attribute "mtime". This then does not allow to have one file list with two time based sort columns.

@PVince81 Can we split this sorting from "data-sort" into "data-sort-type" (by time) and "data-sort-attribute" (expiration time, share time, ...)?

PVince81 commented:
@MorrisJobke feel free

jancborchardt commented:
One high-level issue which is bugging me here is that the left navigation is just a filter. But in this case we show different columns, instead of modified date and size. Do we really want to do that?

PVince81 commented:
The attribute `data-expiration` says `NaN`, please make that either empty string or remove the attribute.

PVince81 commented:
Looks good for now :+1: 

Something for later: re-render the file list row whenever share info changed, because the expiration should update there too. This is a bit more complicated and might influence/break/obsolete the part that updates the share icon.

PVince81 commented:
@MorrisJobke can you fix this https://github.com/owncloud/core/pull/19513#issuecomment-148089667 ?

Then this will need a second reviewer @rullzer @schiesbn

rullzer commented:
@MorrisJobke can you fix https://github.com/owncloud/core/pull/19513#issuecomment-148089667 and rebase?

LukasReschke commented:
Conflicts with master => Moving back to 2

MorrisJobke committed:
Not set expiration date is represented as 0

MorrisJobke commented:
> @MorrisJobke can you fix #19513 (comment) and rebase?

Both done. It still works here :)

MorrisJobke commented:
@PVince81 @LukasReschke @rullzer @Xenopathic Please review :)

rullzer commented:
:+1:

karlitschek commented:
unit tests needed.

LukasReschke commented:
Also displayed at "shared with others". Needs to be fixed:

![2016-03-31_19-48-14](https://cloud.githubusercontent.com/assets/878997/14185305/8bfef126-f779-11e5-99ed-c982d017b15b.png)

MorrisJobke committed:
Update header during update callback

MorrisJobke commented:
> Also displayed at "shared with others". Needs to be fixed:

Should be work now :)

rullzer commented:
@MorrisJobke see https://github.com/owncloud/core/pull/19513#issuecomment-204048259 ;)

MorrisJobke commented:
@rullzer Mind to help me with the unit tests? 🐱

PVince81 commented:
We're past feature-freeze, moving this addition to 9.2. Hopefully there will be time to adjust the unit tests.

MorrisJobke committed:
add unit tests

MorrisJobke commented:
I just added the unit tests ;)

LukasReschke commented:
Works 👍

PVince81 commented on apps/files_sharing/js/sharedfilelist.js:
```diff
@@ -79,6 +79,42 @@
+					text = OC.Util.relativeModifiedDate(expirationTimestamp);
+				} else {
+					formatted = t('files_sharing', 'No expiration date set');
+					text = t('files_sharing', 'Never');
```
@MorrisJobke make it empty as per @jancborchardt's suggestion above ?

schiessle replied:
good point, this also makes it easier to spot the real existing expire dates

PVince81 commented on apps/files_sharing/tests/js/sharedfilelistSpec.js:
```diff
@@ -512,6 +515,15 @@ describe('OCA.Sharing.FileList tests', function() {
 			fileList.reload();
 
+			var currentdate = new Date();
+			var expirationDateInADay =
```
I thought the `moment` library should help to avoid such gymnasitcs

mar1u5 commented:
Nice :+1:

MorrisJobke committed:
Simplify test code and don't show anything if no date is set

MorrisJobke commented:
@PVince81 @schiessle Your comments are addressed :)

PVince81 commented:
Looks good :+1:

JSUnit fail unrelated (fixed separately on master)
